### PR TITLE
fix: add slack_api_url_file to the alertmanager types

### LINKED ISSUE #4045: slack_api_url_file value is missing
**What happened?**

**Did you expect to see something different?**

**How to reproduce it (as minimally and precisely as possible)**:

Use config from here: 
https://github.com/prometheus/alertmanager/commit/59c7fd5053d503037702ad10e4a733aac89e615a

* Prometheus Operator Logs:

```
failed: provision alertmanager configuration: base config from Secret could not be parsed: yaml: unmarshal errors:\n  line 2: field slack_api_url_file not found in type config.plain"
```

**Anything else we need to know?**:

the file to change is : https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/alertmanager/types.go

The code changes are: 

line 54 : SlackAPIURLFile  string          `yaml:"slack_api_url_file,omitempty" json:"slack_api_url_file,omitempty"`

line158: 	APIURLFile    string            `yaml:"api_url_file,omitempty" json:"api_url_file,omitempty"`

### PR DESCRIPTION
**Pull request by djamaile**
so that the operator can provision the config of the alertmanager

Fixes #4045

**Release Note Template (will be copied)**

```
add slack_api_url_file variable to the alertmanager config type
```

### CODE REVIEW

File: pkg/alertmanager/types.go [hunk 1]
```diff
@@ -51,6 +51,7 @@ type globalConfig struct {
 	SMTPAuthIdentity string          `yaml:"smtp_auth_identity,omitempty" json:"smtp_auth_identity,omitempty"`
 	SMTPRequireTLS   *bool           `yaml:"smtp_require_tls,omitempty" json:"smtp_require_tls,omitempty"`
 	SlackAPIURL      *config.URL     `yaml:"slack_api_url,omitempty" json:"slack_api_url,omitempty"`
+	SlackAPIURLFile  *config.URL     `yaml:"slack_api_url_file,omitempty" json:"slack_api_url_file,omitempty"`
```
Comment (simonpasquier) [inline pkg/alertmanager/types.go:54]:
```suggestion
	SlackAPIURLFile  string     `yaml:"slack_api_url_file,omitempty" json:"slack_api_url_file,omitempty"`
```

File: pkg/alertmanager/types.go [hunk 2]
```diff
@@ -154,6 +155,7 @@ type slackConfig struct {
 	VSendResolved *bool             `yaml:"send_resolved,omitempty" json:"send_resolved,omitempty"`
 	HTTPConfig    *httpClientConfig `yaml:"http_config,omitempty" json:"http_config,omitempty"`
 	APIURL        string            `yaml:"api_url,omitempty" json:"api_url,omitempty"`
```
Comment (simonpasquier) [inline pkg/alertmanager/types.go:157]: While we are at it:

```suggestion
	APIURL        *config.URL            `yaml:"api_url,omitempty" json:"api_url,omitempty"`
```

**Review by simonpasquier — COMMENTED** [review]
`slack_api_url_file` and `api_url_file` have been introduced in v0.22.0, the operator should probably reject the config (or stomp the fields?) if `AlertmanagerSpec.Version` is less than v0.22.0.

**Comment by djamaile** [comment]
> `slack_api_url_file` and `api_url_file` have been introduced in v0.22.0, the operator should probably reject the config (or stomp the fields?) if `AlertmanagerSpec.Version` is less than v0.22.0.

So, you want me to build in a condition to reject the config that have the slack parameters for versions that don't support it?

**Comment by djamaile** [comment]
Alright I can continue tomorrow, for the version check should I do it in the [checkSlackConfigs](https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/alertmanager/operator.go#L1130) or before the function gets called? If you know a better place let me know.

**Comment by pharaujo** [comment]
This patch does not compile; the change of  the `APIURL` type to `*config.URL` breaks where it is assumed it should be a string.

I haven't tested this out, but it needs something like:

```diff
diff --git a/pkg/alertmanager/amcfg.go b/pkg/alertmanager/amcfg.go
index 9f057bf8..35d177bf 100644
--- a/pkg/alertmanager/amcfg.go
+++ b/pkg/alertmanager/amcfg.go
@@ -352,7 +352,11 @@ func (cg *configGenerator) convertSlackConfig(ctx context.Context, in monitoring
                if err != nil {
                        return nil, errors.Errorf("failed to get key %q from secret %q", in.APIURL.Key, in.APIURL.Name)
                }
-               out.APIURL = strings.TrimSpace(url)
+               apiurl := &config.URL{}
+               if err := apiurl.UnmarshalJSON([]byte(strings.TrimSpace(url))); err != nil {
+                       return nil, err
+               }
+               out.APIURL = apiurl
        }

        var actions []slackAction
```

**Comment by anderius** [comment]
@djamaile Any progress on this? I am waiting for the fix, so your effort is greatly appreciated! :-)

**Comment by PhilipGough** [comment]
@djamaile this feature has been merged to master so feel free to close this

**Pull request closed without merging**